### ISSUE TITLE
dockerfile: ability to specify multiple Dockerfiles in the same (sub)directory

### ISSUE THREAD

**Issue by dreamcat4** [michaelsauter/crane#155]
Have been talking on `#docker` IRC about this today. Some people need to have 'variant' images. Where the Dockerfile is nearly the same (tweaked a bit). So then it's neater to share the context (support files) and have multiple Dockerfiles in the same directory.

The syntax needed looks like this:

```sh
docker build -f dir/Dockerfile.2 dir

# where dockerfile: dir/Dockerfile.2
```

rather than

```sh
docker build dir

# where dockerfile: dir
```

Reading from `man docker-build` this can be done with the existing `dockerfile:` configuration key if testing for file or directory. If `dockerfile:` resolves to being a directory (e.g. `test -d`) then use the 2nd command. As crane currently does always.

If `dockerfile:` resolves to being a file (e.g. `test -f`). Then get the directors components (everything up until the last `/`). Or use `.` for `dir` if there are no slashes in the value of `dockerfile:`. Which is a corner case. Then use the first above command with `-f` flag ^^.

Sorry if that explanation isn't terribly clear. Here is the place where the source code is running the `docker build` command:

https://github.com/michaelsauter/crane/blob/master/crane/container.go#L732-740

**Comment by michaelsauter** [comment]
I think I got it :)

Makes sense to me, and would be possible to introduce without BC breaks, so I'm :+1: 

I'm in the middle of moving overseas, so no time at all to implement this right now. Do you want to give it a shot?

**Comment by dreamcat4** [comment]
Don't quite have the time either ATM. But sure - if (more likely when) I end up needing it in the future. Unless someone else gets there first.

**Comment by dreamcat4** [comment]
Right. Well it seems real life is not so simple as we might hope.

Due to URLs being also a valid build source. After another look at this. And going inside the actual source code of the `docker build` command and docker API. We cannot fully support docker build options (the command line OR build/ api) in the originally proposed way.

@michaelsauter So I will need you to decide upon a new variable name for the `PATH | URL` mandatory build argument. Remembering that it sets the docker build context. You have 2 best choose from: Either `context:` or `location:`. Please choose one of them.

This new approach also helps us to properlay support any future functionality that may get added to the `docker build` command.

The new key for setting the build context can take over that role previously fulfilled by `dockerfile:`. Having the same meaning as was before ascribed. That will be cleaner (solves URL and other build sources problem which are not local files), and is going to be less bug prone / no inconsistencies with official ways. The new key for specifying the build context will augment the previous function, which will still work 100% for backwards compatibility for old ways.

The variable name `dockerfile` is a part of the official docker build/ api endpoint and set by the optional `--file=` flag on the command line. So it will now be possible to use our `dockerfile:` key in crane to set an alternative `Dockerfile` located inside of the specified build context (if set).

I will also submit here (in same commits) the necessary documentation to explain new behaviour. Including a notice for backwards compatibility about the old way. E.g. if the build context is not set, then `dockerfile` is used as the build context. Or else the folder where the dockerfile is located at.

It will all make more sense when I eventually submit with the ammendment to the crane's READMNE documentation. For now don't be too concerned just please express your preference for the new key name, which may be called either `location:` or `context:`. Many thanks.

**Comment by dreamcat4** [comment]
@michaelsauter But also I've been meaning to ask you Michael:

If you are busier now with other commitments to give crane your full attention. Then why not offer maintainership / collaboration to @bjaglin for a while and see how it goes? He could take over the issue tracker, peer review and merge other people's contributions. I would be more than happy for @bjaglin to look at my code. He is seems to be really trustworthy person, who is level-headend and already demonstrated he is competent enough to do such tasks. And you could still have finally say on those areas you really feel inclined to speak up about. Or are deferred over to you by @bjaglin that he is unsure about or cannot do himself. What do you think?

Would be a much happier to see your official approval & help ensure that crane can be relied upon in future. And that it does not stay as some 1-man show. Rather that, than to end up sending whatever un-merged pull requests over to @bjaglin's fork, i.e. after no response here or behind your back. Or be forced into creating separate unofficial fork / split or whatever. Which may leave out other changes and improvements which everybody wants. To avoid new good stuff going on in parallel, drifting away, and getting left out forever b/c it would be too much work to re-merge. We have to stay on top of these things! Not that my own track record is any better than yours @michaelsauter.

**Comment by dreamcat4** [comment]
There is too few of us - we gotta stick together!

**Comment by dreamcat4** [comment]
Regarding the new key name choosing between: `location:` or `context:`. I've decided to make in my commit with the new key named as `location:` Because that one is the more user-friendly and straightforward term of those two. In other words - it sounds better, less awkward / hard to explain.

Have just done a due diligence, and checked that naming choice is unlikely to conflict or be mis-interpreted to any other known possible feature (which are not implemented yet).

**Comment by michaelsauter** [comment]
@dreamcat4 Yes, I will need to look into how the future maintenance of Crane will happen. I'm going to do that in a couple of weeks.

That being said, I do want to keep Crane as small as possible because I believe that is the best for most people. Which means I will continue to be very selective about the features being merged.

Regarding this specific issue, I will need to have a look which I'm going to do in a few days/weeks.

**Comment by dreamcat4** [comment]
@michaelsauter I Have just spend last 2 days implementing it. So it is already done now. As you gave me a go-ahead to do it. Just finishing things up. There will be a pull request ready later today.

It maintains all of the backwards compatibility (like you asked for).

**Comment by dreamcat4** [comment]
OK @michaelsauter. The pull request is submitted. It has only 1 commit. Have all tested and documented. The code is in good condition, ready for inspection.

Actually there was going to be a more of it. A tarball handling part which I have now taken out. Since your request to keep things as small as possible. As tarball support would have required extra libs (for detecting the tarball, piping to stdin etc.). Increasing the binary's file size from 5.2MB to at least 5.9MB. So it didn't seem worth it in the end. Please be rest assured, I have considered your latest comments too, making these the minimum changes / minimum code possible to implement the given new functionality provided (whilst being bug-free).

Many thanks.

**Comment by michaelsauter** [comment]
This will become obsolete in Crane 2, which has a `build` key. Right now, that only supports `context` (which is a drop-in replacement for the `dockerfile` key) but in the future we can add `file` etc. See #225.

Eventually, what your discussing here will be supported after all :)

**Comment by dreamcat4** [comment]
Great! I'm all for it.
